Tools/LocomotionTool/LocomotionTool.cs

```diff
@@ -0,0 +1,559 @@
+#if UNITY_EDITOR
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR.Core;
+using UnityEditor.Experimental.EditorVR.Utilities;
+using UnityEngine;
+using UnityEngine.InputNew;
+using UnityEngine.UI;
+
+namespace UnityEditor.Experimental.EditorVR.Tools
+{
+	sealed class LocomotionTool : MonoBehaviour, ITool, ILocomotor, IUsesRayOrigin, ISetDefaultRayVisibility, ICustomActionMap,
+		ILinkedObject, IUsesViewerScale, ISettingsMenuItemProvider, ISerializePreferences
+	{
+		const float k_FastMoveSpeed = 20f;
+		const float k_SlowMoveSpeed = 1f;
+		const float k_RotationDamping = 0.2f;
+		const float k_RotationThreshold = 0.75f;
+
+		//TODO: Fix triangle intersection test at tiny scales, so this can go back to 0.01
+		const float k_MinScale = 0.1f;
+		const float k_MaxScale = 1000f;
+
+		const string k_WorldScaleProperty = "_WorldScale";
+
+		const int k_RotationSegments = 32;
+
+		enum State
+		{
+			Inactive = 0,
+			Aiming = 1,
+			Moving = 3
+		}
+
+		[SerializeField]
+		GameObject m_BlinkVisualsPrefab;
+
+		[SerializeField]
+		GameObject m_ViewerScaleVisualsPrefab;
+
+		[SerializeField]
+		ActionMap m_BlinkActionMap;
+
+		[SerializeField]
+		GameObject m_SettingsMenuItemPrefab;
+
+		[Serializable]
+		class Preferences
+		{
+			[SerializeField]
+			bool m_BlinkMode;
+
+			public bool blinkMode { get { return m_BlinkMode; } set { m_BlinkMode = value; } }
+		}
+
+		Preferences m_Preferences;
+
+		ViewerScaleVisuals m_ViewerScaleVisuals;
+
+		GameObject m_BlinkVisualsGO;
+		BlinkVisuals m_BlinkVisuals;
+
+		State m_State = State.Inactive;
+
+		bool m_AllowScaling = true;
```

Comment:
Yeah. A big improvement in this iteration is splitting out the various types of locomotion into their own method. In fact, the reason this was originally BlinkLocomotionTool is that we had intended for users to explicitly swap different locomotion tools for different types, but then realized that we pretty much want everything on all the time, if possible.